### DISCUSSION TITLE
Nested Joins with dart

### DISCUSSION THREAD

**Discussion by lifeapps-42** [simolus3/drift#2232] [Q&A] [answered]
FOA, thank you for the great and huge work!

I know, this is a frequent question, but I still struggle to do nested joins. I'm not very experienced with sql, so maybe i'm just getting smth wrong.

So, here's simplified setup:

I have 3 tables:

Messages:
| id | text | chat_room | user |
|---|---|---|---|
| int | text | int | int -> Users.id |

Users:
| id | avatar | name |
|---|---|---|
| int | int -> MedaiObjects.id | text |

MediaObjects:
| id | preview | content |
|---|---|---|
| int | text | text |

The goal is to compose object from them quering by `chat_room` id.

It is pretty clear how to join `User` entity with `Message`. The problem is to fetch `MediaObject` entity since at a moment of join I do not know what the media object id will be. 

Did i miss any similar discussion? 

Thank you for any help!

**Comment by simolus3** [comment, marked as answer]
> It is pretty clear how to join User entity with Message

If you have a `Messages` table in a select, you could also join the media object for the user's avatar:

```dart
join([
  innerJoin(users, on: users.id.equalsExp(messages.user)),
  innerJoin(mediaObjects, on: mediaObjects.id.equalsExp(users.avatar)),
])
```

But I guess you're joining/querying the users in a different way that doesn't make the messages table available directly? If you could post the code you've used in "join `User` entity with `Message`", I might be able to give clearer advice :)

**Reply by lifeapps-42** [reply]
Ok, I couldn't just fit ithis in mind)) What i was trying to do is more like that:

```dart
join([
  innerJoin(users, on: users.id.equalsExp(messages.user)),
]).join([
   innerJoin(mediaObjects, on: mediaObjects.id.equalsExp(users.avatar)),
  ]);
```

:) now i understood) Thanks a lot

And, some quick offtopic, if you don't mind.

Since I need this User join quite a lot, could it be better solution to use View for reusing? Ex: To get full Messege entity can I just join the MessagesTable with the UsersView? Or more of that: Can i create the MessageView referencing the MessageTable joined with UserView? Has this view nesting strategy some performance drawbacks against simple joining?

Best wishes an Happy Christmas!

**Reply by simolus3** [reply]
You could also define a view for this, yes. I can't think of any obvious performance advantages/disadvantages for either approach.

> Can i create the MessageView referencing the MessageTable joined with UserView

You can, there's [an example](https://drift.simonbinder.eu/docs/getting-started/advanced_dart_tables/#views) for a view with a join in the documentation. However, be aware that views don't currently allow you to read tables back (there's no `readTable` for views). Instead, drift generates a data class for this view, but this data class will then have all the combined columns. If that works for you, it's a good way to make the common join easier to write. It's just something to be aware of :)

**Reply by lifeapps-42** [reply]
Thanks a lot!
